statesync/reactor.go

```diff
@@ -0,0 +1,274 @@
+func (r *Reactor) Receive(chID byte, src p2p.Peer, msgBytes []byte) {
+	if !r.IsRunning() {
+		return
+	}
+
+	msg, err := decodeMsg(msgBytes)
+	if err != nil {
+		r.Logger.Error("Error decoding message", "src", src, "chId", chID, "msg", msg, "err", err, "bytes", msgBytes)
+		r.Switch.StopPeerForError(src, err)
+		return
+	}
+	err = msg.ValidateBasic()
+	if err != nil {
+		r.Logger.Error("Invalid message", "peer", src, "msg", msg, "err", err)
+		r.Switch.StopPeerForError(src, err)
+		return
+	}
+
+	switch chID {
+	case SnapshotChannel:
+		switch msg := msg.(type) {
+		case *snapshotsRequestMessage:
+			snapshots, err := r.recentSnapshots(recentSnapshots)
+			if err != nil {
+				r.Logger.Error("Failed to fetch snapshots", "err", err)
+				return
+			}
+			for _, snapshot := range snapshots {
+				r.Logger.Debug("Advertising snapshot", "height", snapshot.Height,
+					"format", snapshot.Format, "peer", src.ID())
+				src.Send(chID, cdc.MustMarshalBinaryBare(&snapshotsResponseMessage{
+					Height:      snapshot.Height,
+					Format:      snapshot.Format,
+					ChunkHashes: snapshot.ChunkHashes,
+					Metadata:    snapshot.Metadata,
+				}))
+			}
+
+		case *snapshotsResponseMessage:
+			r.mtx.RLock()
+			defer r.mtx.RUnlock()
+			if r.syncer == nil {
+				r.Logger.Debug("Received unexpected snapshot, no state sync in progress")
+				return
+			}
+			r.Logger.Debug("Received snapshot", "height", msg.Height, "format", msg.Format, "peer", src.ID())
+			_, err := r.syncer.AddSnapshot(src, &snapshot{
+				Height:      msg.Height,
+				Format:      msg.Format,
+				ChunkHashes: msg.ChunkHashes,
+				Metadata:    msg.Metadata,
+			})
+			if err != nil {
+				r.Logger.Error("Failed to add snapshot to sync", "height", msg.Height,
+					"format", msg.Format, "peer", src.ID(), "err", err)
+				return
+			}
+
+		default:
+			r.Logger.Error("Received unknown message %T", msg)
+		}
+
+	case ChunkChannel:
+		switch msg := msg.(type) {
+		case *chunkRequestMessage:
+			r.Logger.Debug("Received chunk request", "height", msg.Height, "format", msg.Format,
+				"chunk", msg.Chunk, "peer", src.ID())
+			resp, err := r.conn.LoadSnapshotChunkSync(abci.RequestLoadSnapshotChunk{
+				Height: msg.Height,
+				Format: msg.Format,
+				Chunk:  msg.Chunk,
+			})
+			if err != nil {
+				r.Logger.Error("Failed to load chunk", "height", msg.Height, "format", msg.Format,
+					"chunk", msg.Chunk, "err", err)
+				return
+			}
+			r.Logger.Debug("Sending chunk", "height", msg.Height, "format", msg.Format,
+				"chunk", msg.Chunk, "peer", src.ID())
+			src.Send(ChunkChannel, cdc.MustMarshalBinaryBare(&chunkResponseMessage{
+				Height:  msg.Height,
+				Format:  msg.Format,
+				Chunk:   msg.Chunk,
+				Body:    resp.Chunk,
+				Missing: resp.Chunk == nil,
+			}))
+
+		case *chunkResponseMessage:
+			r.mtx.RLock()
+			defer r.mtx.RUnlock()
+			if r.syncer == nil {
+				r.Logger.Debug("Received unexpected chunk, no state sync in progress", "peer", src.ID())
```

Comment:
how's this possible? aren't syncer always present when we're in state sync mode? otherwise, we shut down reactor when it's not needed, right?

Reply:
No. The reactor also serves snapshots for other nodes that may want to run state sync. So the reactor will always run, but will rarely do state sync (only on bootstrap).